jaxrs-api/src/main/java/javax/ws/rs/JAXRS.java

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright (c) 2018 Markus KARG. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0, which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+
+package javax.ws.rs;
+
+import javax.ws.rs.core.Application;
+import javax.ws.rs.ext.RuntimeDelegate;
+
+/**
+ * Bootstrap class that is used to startup a JAX-RS application in Java SE
+ * environments.
+ * <p>
+ * The {@code JAXRS} class is available in a Jakarta EE container environment as
+ * well; however, support for the Java SE bootstrapping APIs is <em>not
+ * required</em> in container environments.
+ * </p>
+ * <p>
+ * In a Java SE environment an application is getting started by the following
+ * command using default configuration values (i. e. mounting application at
+ * {@code http://localhost:80/}:
+ * </p>
+ *
+ * <pre>
+ * JAXRS.start(new MyApplication(), JAXRS.Configuration.builder().build());
+ * </pre>
+ * <p>
+ * The following example shows how to override defaut values for protocol, host,
+ * port and root path:
+ * </p>
+ *
+ * <pre>
+ * JAXRS.Configuration config = JAXRS.Configuration.builder().protocol("HTTP").host("0.0.0.0").port(80).rootPath("api")
+ * 	.build();
+ * JAXRS.start(new MyApplication(), config);
+ * </pre>
+ *
+ * @author Markus KARG (markus@headcrashing.eu)
+ *
+ * @since 2.2
+ */
+public interface JAXRS {
+
+    /**
+     * Invoked in Java SE environments to start the provided application at the
+     * specified root URL.
+     *
+     * This method will not return until the JAX-RS application is terminated.
```

Comment:
The new proposal looks much better Markus. Here are a few comments:

- I believe start/stop should be async so that apps can do other inits during startup on main thread. I would use CompletionStage<Instance> as return value. That way you can also do,

 JAXRS.start(app, conf).thenAccept(instance -> ...)

- I'm not sold on the name "Instance". Is "Server" better?

Reply:
I see your point and in fact had the same vision first, but then I abstained from it, because:
* If I am writing any Java SE program that shall do multiple things at the same time, it is my job to provide the needed parallelization plumbing, as there are virtually no APIs in the Java SE world that return CompletableStages as of today, and I simply could not find a good argument why JAX-RS is so special that it should provide a CompletionStage as the first and only one. If someone wants continuations, he can simply wrap the synchronous start() method by CompletableFuture.supplyAsync. So the sole benefit would be that a JAX-RS implementation would not need to synthetically wait for the end of the bootsrap. That means OTOH if the majority of existing implementations would be async, certainly it would make sense to simply disclose the internal stage.
* "Server" came in to my mind first, but isn't it a bit too broad? I think it would be not concise enough, as with "Server" people might imply "Java EE Server", "Web Server", "Servlet Server", etc. It is just an overstressed term. "Instance" (in the sense of "Application Instance") seems to be more accurate, as in fact, the Java SE programmer does not start a Server, he starts a local application instance.

Anyways, I have no strong feelings about the terms. Committers, what do you think? We simply could vote. Here is mine:

async: 0
Instance: +1
Server: 0

Reply:
Since nobody else is providing feedback, here are some additional comments ;)

(1) I don't the follow your argument about about CompletionStage. Lots of APIs use it, in fact we use it in JAX-RS already :) This is about your application's main, which will likely need to do other things in addition to starting JAX-RS. Given that the start method is non-trivial, it makes perfect sense to make it async in this context.

(2) I feel that CDI support is such an important use case that I'm wondering whether the builder should support it. I realize this is somewhat orthogonal to JAX-RS, but practically, very useful if I can start my JAX-RS server with CDI enabled.

Reply:
(1) What I meant to say is that Java SE 8 (= the sole target of my proposal) does not contain other APIs that produce CompletableFuture, so ALL other work besides starting JAX-RS has to be manually wrapped by the application programmer anyways. Hence there is no benefit. We just spare him one sole line of code. Anyways, I am not against it.

(2) CDI is optional in JAX-RS 2.1 and not mentioned in Java SE 8 at all, but it might be a good idea to simply add a method like `.cdi(SeContainer javaseCdiContainer)` to pass the Java SE 8 CDI container initialized by the `main()` method of the standalone bootstrap code? The JAX-RS implementations can then further configure that container.

Reply:
(1) I like the CompletionStage idea and the start being asynchronous, but .. on which thread it would run? Shouldn't we have a threadpools-related method in Configuration class then? (it's somehow similar to RxInvokerProvider in this context) 

(2) I believe the CDI bootstrap routine would be little more complex than that and single builder method might not be enough.

From what I recall, CDI does require all classes / providers to be registered before the container is initializer (ready to serve beans). That might actually require completely different kind of API if there is a need to make it part of the "jax-rs" initializer.

(Obviously, that depends on how the CDI integration would be defined for this particular usecase, but I believe it should be possible to inject any application-scoped instances which JAX-RS runtime creates to any other "app-scoped" bean).

Reply:
I leave it up to @spericas to answer on that, as it was *him* who wanted both. I just reacted upon his ideas and can live without `CompletionStage` *and* CDI. Again, we do this for 2.2, which does *not* mandate CDI.

OTOH if it is really impossible to add classes / providers to an existing `SeContainer` then we could simply do `.cdi(ContainerInitializer)` which allows that. But before we discuss the actual solution it might be good to ask someone who is really deep into Java SE CDI initialization... @struberg Can you please help us?

Reply:
Hi @pavelbucek. On (1) we can add a method to config and possibly use ForkJoinPool.commonPool() by default. I'm not sure I follow the issue on (2), are you thinking about a JAX-RS/CDI issue or a Jersey/CDI issue?